src/inventory_ui.cpp

```diff
@@ -108,308 +393,413 @@ bool inventory_column::is_selected( const inventory_entry &entry ) const
+            match->push_back( &*it );
         } else {
-            res.emplace_back( 1, const_cast<item *>( &*it ) );
+            res.emplace_back( 1, &*it );
         }
     }
 
     return res;
 }
 
-void inventory_selector::add_custom_items( const std::list<item>::const_iterator &from,
-                                           const std::list<item>::const_iterator &to,
-                                           const std::string &title,
+const item_category *inventory_selector::naturalize_category( const item_category &category, const tripoint &pos )
+{
+    const auto find_cat_by_id = [ this ]( const std::string &id ) {
+        const auto iter = std::find_if( categories.begin(), categories.end(), [ &id ]( const item_category &cat ) {
+            return cat.id == id;
+        } );
+        return iter != categories.end() ? &*iter : nullptr;
+    };
+
+    const int dist = rl_dist( p.pos(), pos );
+
+    if( dist != 0 ) {
+        const std::string suffix = direction_suffix( p.pos(), pos );
+        const std::string id = string_format( "%s_%s", category.id.c_str(), suffix.c_str() );
+
+        const auto existing = find_cat_by_id( id );
+        if( existing != nullptr ) {
+            return existing;
+        }
+
+        const std::string name = string_format( "%s %s", category.name.c_str(), suffix.c_str() );
+        const int sort_rank = category.sort_rank + dist;
+        const item_category new_category( id, name, sort_rank );
+
+        categories.push_back( std::move( new_category ) );
+    } else {
+        const auto existing = find_cat_by_id( category.id );
+        if( existing != nullptr ) {
+            return existing;
+        }
+
+        categories.push_back( category );
+    }
+
+    return &categories.back();
+}
+
+void inventory_selector::add_item( inventory_column &target_column,
+                                   const std::shared_ptr<item_location> &location,
+                                   size_t stack_size, const item_category *custom_category )
+{
+    if( !preset.is_shown( *location ) ) {
+        return;
+    }
+
+    inventory_entry entry( location, stack_size, custom_category );
+
+    entry.enabled = preset.is_enabled( entry.get_location() );
+    entry.rank = preset.get_rank( entry.get_location() );
+
+    target_column.add_entry( entry );
+    on_entry_add( entry );
+
+    layout_is_valid = false;
+}
+
+void inventory_selector::add_custom_items( inventory_column &target_column,
+                                           const std::list<item>::iterator &from,
+                                           const std::list<item>::iterator &to,
+                                           const item_category &custom_category,
                                            const std::function<std::shared_ptr<item_location>( item * )> &locator )
 {
     const auto &stacks = restack_items( from, to );
+    const item_category *nat_category = nullptr;
 
     for( const auto &stack : stacks ) {
```

Comment:
Variables that differ only by one name are evil. You could consider using `e` or `sk` for the iteration variable